libs/utils/src/sync/heavier_once_cell.rs

```diff
@@ -0,0 +1,306 @@
+use std::sync::{Arc, Mutex, MutexGuard};
+use tokio::sync::Semaphore;
+
+/// Custom design like [`tokio::sync::OnceCell`] but using [`OwnedSemaphorePermit`] instead of
+/// `SemaphorePermit`, allowing use of `take` which does not require holding an outer mutex guard
+/// for the duration of initialization.
+///
+/// Has no unsafe, builds upon [`tokio::sync::Semaphore`] and [`std::sync::Mutex`].
+///
+/// [`OwnedSemaphorePermit`]: tokio::sync::OwnedSemaphorePermit
+pub struct OnceCell<T> {
+    inner: Mutex<Inner<T>>,
+}
+
+impl<T> Default for OnceCell<T> {
+    /// Create new uninitialized [`OnceCell`].
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+        }
+    }
+}
+
+/// Semaphore is the current state:
+/// - open semaphore means the value is `None`, not yet initialized
+/// - closed semaphore means the value has been initialized
+#[derive(Debug)]
+struct Inner<T> {
+    init_semaphore: Arc<Semaphore>,
+    value: Option<T>,
+}
```

Comment:
How is this different from

```rust
struct Inner {
    value: Arc<RwLock<Option<T>>>
}
```

Reply:
It's not an rwlock. You'd have to either to do racy-init (first one wins, other's get dropped, which would be quite bad worst case for 2download and init"), or .. Well, I don't know how else you'd initialize this..

I explained above how this allows for non-lock holding asynchronous initialization: https://github.com/neondatabase/neon/pull/4938#discussion_r1307770922 (as would `tokio::sync::OnceCell` allow, but unlike `Mutex<tokio::sync::OnceCell>`, we can do asynchronous drop -- the eviction -- on this value while holding the permit and thus forbidding re-initialization while this "asynchronous drop" is ongoing).

I called the non-lock holding initialization as "unsynchronized" in earlier revisions. It was a bad idea.